gossip/minimal/mysql/hawkstate.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mysql provides a MySQL based implementation of persistent
+// state management for the goshawk tool.
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/google/certificate-transparency-go/gossip/minimal"
+)
+
+func NewStateManager(ctx context.Context, db *sql.DB) (minimal.ScanStateManager, error) {
+	tx, err := db.BeginTx(ctx, nil /* opts */)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create state transaction: %v", err)
+	}
+	defer tx.Commit()
+	rows, err := tx.QueryContext(ctx, "SELECT s.HubURL, s.NextIndex FROM HubNext AS s;")
+	if err != nil {
+		return nil, fmt.Errorf("failed to query state rows: %v", err)
+	}
+	defer rows.Close()
+	m := mysqlStateManager{
+		ScanState: minimal.ScanState{Next: make(map[string]int64)},
+		db:        db,
+	}
+	glog.Info("Reading scan state from DB")
+	for rows.Next() {
+		var hubURL string
+		var nextIndex int64
+		if err := rows.Scan(&hubURL, &nextIndex); err != nil {
+			return nil, fmt.Errorf("failed to scan state row: %v", err)
+		}
+		glog.Infof("  scanState[%q]=%d", hubURL, nextIndex)
+		m.ScanState.Next[hubURL] = nextIndex
+	}
+	return &m, nil
+}
+
+type mysqlStateManager struct {
+	minimal.ScanState
+	db *sql.DB
+}
+
+func (m *mysqlStateManager) Flush(ctx context.Context) error {
+	m.Mu.Lock()
+	defer m.Mu.Unlock()
+
+	tx, err := m.db.BeginTx(ctx, nil /* opts */)
+	if err != nil {
+		return fmt.Errorf("failed to create state transaction: %v", err)
+	}
+	defer tx.Commit()
+
+	glog.Info("Flushing scan state to DB")
+	for url, index := range m.ScanState.Next {
+		glog.Infof("  scanState[%q]=%d", url, index)
+		_, err = tx.ExecContext(ctx, "REPLACE INTO HubNext(HubURL, NextIndex) VALUES (?, ?);", url, index)
```

Comment:
This sort of bypasses the primary key constraint as it would work if there were multiple copies of the same key. However, the source is a map so there can't be duplicate keys in practice.

Reply:
I think the implicit keep-last behaviour is OK given that the source of the data is a Go `map`.